File: Anagram/scala/src/main/scala/net/silve/dojo/Anagram.scala
package net.silve.dojo

import java.text.Normalizer

object Anagram {

  def apply(list: Seq[String]): Boolean = list.map(prepare) match {
    case Nil => false
    case a :: Nil => false
    case a :: b :: Nil => a == b
    case a :: b :: tail => a == b && apply(tail :+ b)
  }

  private def prepare = { x: String =>
    Normalizer.normalize(x, Normalizer.Form.NFKD)
      .toLowerCase
      .replaceAll("[\\p{InCombiningDiacriticalMarks}]", "")
      .replaceAll("\\W", "").toCharArray.sorted.mkString
  }

}

File: Anagram/java/src/main/java/net/silve/dojo/Anagram.java
package net.silve.dojo;


import java.text.Normalizer;
import java.util.Arrays;

public class Anagram {

    public boolean validate(String a, String b) {
        char[] sortedA = prepare(a).toCharArray();
        Arrays.sort(sortedA);
        char[] sortedB = prepare(b).toCharArray();
        Arrays.sort(sortedB);

        return Arrays.equals(sortedA, sortedB);
    }

    private String prepare(String s) {
        return Normalizer.normalize(s, Normalizer.Form.NFKD)
                .toLowerCase()
                .replaceAll("[\\p{InCombiningDiacriticalMarks}]", "")
                .replaceAll("\\W", "");
    }

}
